src/celeritas/io/RootStepWriter.cc

```diff
@@ -0,0 +1,247 @@
+{
+    CELER_EXPECT(root_manager_);
+    this->make_tree();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Set the number of entries (i.e. number of steps) stored in memory before
+ * ROOT flushes the data to disk. Default is ~32MB of compressed data.
+ *
+ * See `SetAutoFlush` in ROOT TTree Class reference for details:
+ * https://root.cern.ch/doc/master/classTTree.html
+ */
+void RootStepWriter::set_auto_flush(long num_entries)
+{
+    CELER_EXPECT(root_manager_);
+    CELER_EXPECT(tstep_tree_);
+    tstep_tree_->SetAutoFlush(num_entries);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Collect step data from each track on each thread id.
+ */
+void RootStepWriter::execute(StateHostRef const& steps)
+{
+#define IF_SELECTED(ATTR, VAL) \
+    do                         \
+    {                          \
+        if (selection_.ATTR)   \
+        {                      \
+            tstep_.ATTR = VAL; \
+        }                      \
+    } while (0)
+
+#define IF_POINT_SELECTED(ATTR, VAL) \
+    do                               \
+    {                                \
+        if (selection_point.ATTR)    \
+        {                            \
+            tpoint.ATTR = VAL;       \
+        }                            \
+    } while (0)
+
+#define IF_POINT_REAL3_SELECTED(ATTR, TPOINT_ATTR)          \
+    do                                                      \
+    {                                                       \
+        if (selection_point.ATTR)                           \
+        {                                                   \
+            this->copy_real3(point.ATTR[tid], TPOINT_ATTR); \
+        }                                                   \
+    } while (0)
+
+    // TODO: add selection_ options for storing data
+    CELER_EXPECT(steps);
+    tstep_ = mctruth::TStepData();
+
+    // Loop over thread ids and fill TTree
+    for (const auto tid : range(ThreadId{steps.size()}))
+    {
+        if (!steps.track_id[tid])
+        {
+            // Track id not found; skip inactive track slot
+            continue;
+        }
+
+        // Track id is always set
+        tstep_.track_id = steps.track_id[tid].unchecked_get();
+
+        IF_SELECTED(event_id, steps.event_id[tid].get());
+        IF_SELECTED(action_id, steps.action_id[tid].get());
+        IF_SELECTED(particle, particles_->id_to_pdg(steps.particle[tid]).get());
+        IF_SELECTED(energy_deposition, steps.energy_deposition[tid].value());
+        IF_SELECTED(step_length, steps.step_length[tid]);
+        IF_SELECTED(track_step_count, steps.track_step_count[tid]);
+
+        // Store pre- and post-step
+        for (auto i : range(StepPoint::size_))
+        {
+            const auto& selection_point = selection_.points[i];
+            const auto& point           = steps.points[i];
+            auto&       tpoint          = tstep_.points[(int)i];
```

Comment:
If you take my above suggestion this should no longer be necessary, but never use C-style casts like this (yes, I know it's all over the place in ROOT and geant, but it's dangerous because it can accidentally be a reinterpret cast, const cast, or a combination). Instead use `static_cast<int>(i)`.

https://en.cppreference.com/w/cpp/language/explicit_cast